doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,979 @@
+
+.. code:: C++
+
+    #include "llvm/Pass.h"
+    #include "llvm/IR/Function.h"
+    #include "llvm/Support/raw_ostream.h"
+
+    #include "llvm/IR/LegacyPassManager.h"
+    #include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+    using namespace llvm;
+
+    namespace {
+    class MyPass : public BasicBlockPass {
+    public:
+      static char ID;
+
+      MyPass() : BasicBlockPass(ID) {}
+
+      virtual bool runOnBasicBlock(BasicBlock &BB) override {
+        errs() << "I m running on a block...\n";
+        return false;
+      }
+
+    };
+    }
+
+    char MyPass::ID = 0;
+    static RegisterPass<MyPass> X("MyPass", "Obfuscates zeroes",
+                                         false, false);
+
+    // register pass for clang use
+    static void registerMyPassPass(const PassManagerBuilder &,
+                                   PassManagerBase &PM) {
+      PM.add(new MyPass());
+    }
+    static RegisterStandardPasses
+        RegisterMBAPass(PassManagerBuilder::EP_EarlyAsPossible,
+                        registerMyPassPass);
+
+If you have been paying attention so far you should remember that we are going to obfuscate null literals.
+And to do so we will randomly pick two variables reachable from where the replacement occurs.
+So, in order to keep the pass as simple as possible we are going to work at the basic bloc level, this way there will be no reachability problems with the variables we encounter.
+This is why our class derives from the ``BasicBlockPass`` class.
+
+This could be greatly enhanced using `dominators <http://llvm.org/docs/doxygen/html/classllvm_1_1DominatorTree.html>`_, but that's... another story!
+
+.. code:: C++
+
+    class MyPass : public BasicBlockPass
+
+
+Do or do not there is no... test
+********************************
+
+I am sure that your are eager to compile and run this empty pass. Thanks to the files provided in the Quarkslab git it's actually quite easy.
+First you need to tell cmake that your pass should be compiled by adding it in the file ``$PASSDIR/llvm-passes/CMakeList.txt``.
+It should now look like this:
+
+.. code:: cmake
+
+    set(EPONA_LLVM_MODULES
+        ObfuscateZero
+        MyPass
+    )
+
+Now we are going to build the pass:
+
+.. code:: bash
+
+    >$ cd $PASSDIR
+    >$ mkdir build
+    >$ cd build
+    >$ cmake -DLLVM_ROOT=path/to/your/llvm/build ..
+    >$ make
+
+And now let's run our pass with clang. But we need a test file, write the following code somewhere:
+
+.. code:: c
+
+    #include <stdio.h>
+
+    int foo(){return 1;}
+
+    int main() {
+        printf("Hello world\n");
+
+        return 0;
+    }
+
+
+
+And compile it with our awesome pass using:
+
+.. code:: bash
+
+    >$ clang++ -Xclang -load -Xclang $PASSDIR/build/llvm-passes/LLVMMyPass.so path/to/test/file -o /tmp/awesome.out
+
+Or if you already have a file with the LLVM bytecode:
+
+.. code:: bash
+
+  >$ opt -load $PASSDIR/build/llvm-passes/LLVMMyPass.so -MyPass in.ll -o out.bc
```

Comment:
-S allows to have a .ll and readable output. No need for bytecode at this point.